numba/unicode.py

```diff
@@ -493,87 +484,121 @@ def unicode_contains(a, b):
-        raise TypingError(msg)
+def unicode_idx_check_type(ty, name):
+    """Check object belongs to one of specific types
+    ty: type
+        Type of the object
+    name: str
+        Name of the object
+    """
+    thety = ty
+    # if the type is omitted, the concrete type is the value
+    if isinstance(ty, types.Omitted):
+        thety = ty.value
+    # if the type is optional, the concrete type is the captured type
+    elif isinstance(ty, types.Optional):
+        thety = ty.type
+
+    accepted = (types.Integer, types.NoneType)
+    if thety is not None and not isinstance(thety, accepted):
+        raise TypingError('"{}" must be {}, not {}'.format(name, accepted, ty))
 
-    def rfind_impl(s, substr, start=None, end=None):
-        length = len(s)
+
+def generate_finder(find_func):
+    """Generate finder either left or right."""
+    def impl(data, substr, start=None, end=None):
+        length = len(data)
         sub_length = len(substr)
         if start is None:
             start = 0
         if end is None:
             end = length
 
-        # https://github.com/python/cpython/blob/201c8f79450628241574fba940e08107178dc3a5/Objects/unicodeobject.c#L9342-L9354
-        def _adjust_indices(length, start, end):
-            if end > length:
-                end = length
-            if end < 0:
-                end += length
-                if end < 0:
-                    end = 0
-            if start < 0:
-                start += length
-                if start < 0:
-                    start = 0
-
-            return start, end
-
         start, end = _adjust_indices(length, start, end)
         if end - start < sub_length:
             return -1
 
-        if sub_length == 0:
-            return end
+        return find_func(data, substr, start, end)
 
-        for i in range(min(len(s), end) - len(substr), start - 1, -1):
-            if _cmp_region(s, i, substr, 0, len(substr)) == 0:
-                return i
-        return -1
-    return rfind_impl
+    return impl
+
+
+@register_jitable
+def _finder(data, substr, start, end):
+    """Left finder."""
+    if len(substr) == 0:
+        return start
+    for i in range(start, min(len(data), end) - len(substr) + 1):
+        if _cmp_region(data, i, substr, 0, len(substr)) == 0:
+            return i
+    return -1
+
+
+@register_jitable
+def _rfinder(data, substr, start, end):
+    """Right finder."""
+    if len(substr) == 0:
+        return end
+    for i in range(min(len(data), end) - len(substr), start - 1, -1):
+        if _cmp_region(data, i, substr, 0, len(substr)) == 0:
+            return i
+    return -1
+
+
+_find = register_jitable(generate_finder(_finder))
+_rfind = register_jitable(generate_finder(_rfinder))
+
+
+@overload_method(types.UnicodeType, 'find')
+def unicode_find(data, substr, start=None, end=None):
+    """Implements str.find()"""
+    if isinstance(substr, types.UnicodeCharSeq):
```

Comment:
I don't quite understand, what this `if` statement is for. It doesn't appear in the `rfind`  implementation below.

Reply:
Added the same for `rfind`. Additionally added `UnicodeCharSeq.rfind`.

Reply:
o.k. can you explain what it does? Does it recursively call itself or something?

Reply:
It casts `substr` from `UnicodeCharSeq` to `UnicodeType` (`str(substr)`), so `s` and `sub` are of Unicode type and respectively the main algorithm is executed. Basically it can be called `recursive call`. This way was implemented before https://github.com/numba/numba/pull/4866/files#diff-a21787256ad779b4c4c5c99915c71025L500-L509 for `find`. So I also added that for `rfind`.

Reply:
Otherwise test `numba.tests.test_unicode_array.TestUnicodeArray.test_return_find` fails.